### ISSUE TITLE
Enhance `showPosition()` to Accurately Display Error Position by Line Number

### ISSUE THREAD

**Issue by satyajitnayk** [handlebars-lang/handlebars.js#2023]
Ensures that the ----^ indicator accurately reflects the error position on the relevant line, rather than defaulting to the last line:

For more visit: https://github.com/zaach/jison/issues/406

**Comment by jaylinski** [comment]
Can you provide an example on how to reproduce the issue?

**Comment by satyajitnayk** [comment]
In the methods `pastInput()` & `upcomingInput()` we are replacing all newline chars with a space right!
<img width="1063" alt="image" src="https://github.com/handlebars-lang/handlebars.js/assets/32722867/10116564-c724-4c37-a9e1-44c40dd31392">

So when I try to remove them & try to get full error template position I am always getting `---^` at last line of template rather than exact error position.
![image](https://github.com/handlebars-lang/handlebars.js/assets/32722867/152195f8-44f3-478d-9218-933133995e6f)



> **My suggestion is that we can improve error message if we could display exact error template instead of converting template to one line.** 



1. **Current error message thrown:**  - even though template is multi line error displayed converts it to single line.
    `{{#if}    HELLO {{/if}}` 
<img width="1467" alt="image" src="https://github.com/handlebars-lang/handlebars.js/assets/32722867/06fd8b67-8625-4b9c-a1b4-330c2081e607">

2. **Improved error message:** (we can improve error position part to display original template & error position @jaylinski. What's your thought on this?? )
```
    Parse error on line 3:
    {{#if}
-------^
    HELLO
    {{/if}}
Expecting 'CLOSE_RAW_BLOCK', 'CLOSE', 'CLOSE_UNESCAPED', 'OPEN_SEXPR', 'CLOSE_SEXPR', 'ID', 'OPEN_BLOCK_PARAMS', 'STRING', 'NUMBER', 'BOOLEAN', 'UNDEFINED', 'NULL', 'DATA', 'SEP', got 'INVALID'
```

**Comment by satyajitnayk** [comment]
> Makes sense. I updated your issue description. Let me know if that is correct.

adjusted expected error message. It looks correct to me @jaylinski

**Comment by satyajitnayk** [comment]
One question: Can we modify methods `showPosition() `, `pastInput()` ,`upcomingInput()` OR are these functions are from different package altogether @jaylinski ??

**Comment by jaylinski** [comment]
Sadly not, the parsing-logic is part of `jison` (you already opened an issue there).

**Comment by satyajitnayk** [comment]
any thoughts/suggestion how could we improve the error message @jaylinski ?
FYI  `jison` was last released 3 years ago & not maintained actively.

**Comment by jaylinski** [comment]
In order to allow users to implement custom error handlers, one would have to rewrite the current parser.

There are projects that attempt to bring handlebars up-to-date, like https://github.com/nknapp/handlebars-ng. Those are probably a better starting point for such big changes.

**Comment by satyajitnayk** [comment]
I doesn't mean to rewrite whole parser, just send the whole error message as it is from Handlebar(with out truncate/manipulation). Let user to manipulate it as per their need @jaylinski

**Comment by jaylinski** [comment]
Yes, directly exposing the `lexer` object (including the error-message) could probably be done.

**Comment by satyajitnayk** [comment]
Let's do it then @jaylinski

**Comment by jaylinski** [comment]
I currently don't have time to implement this. But I'll be willing to review/accept PRs if somebody improves the error-message (`master`-branch only).

**Comment by satyajitnayk** [comment]
Hi @jaylinski and @nknapp, I've attempted this task but I'm struggling to find a starting point. Could you provide some guidance or a brief overview to help me get started?

**Comment by tiholic** [comment]
@jaylinski @satyajitnayk I'm also looking to address same issue!
